<?xml version="1.0"?>

<launch>
   
  <!-- pddl domain path -->
  <arg name="domain_path"  default="$(find mimree_executive)/pddl/domain.pddl" />
  <arg name="problem_path"  default="" />

  <!-- automatically generated pddl problem will be created here -->
  <arg name="autom_gen_problem_path" default="$(find mimree_executive)/pddl/problem.pddl" />

  <!-- use problem.pddl or receive problem from topic -->
  <arg name="use_problem_topic" default="true" />

  <!-- planner will produce output plan and write to this location -->
  <arg name="data_path" default="$(find mimree_executive)/pddl/" />

  <!-- ROSPlan -->
  <!-- knowledge base -->
  <node name="rosplan_knowledge_base" pkg="rosplan_knowledge_base" type="knowledgeBase" respawn="false" output="screen">
    <param name="domain_path" value="$(arg domain_path)" />
    <param name="problem_path" value="$(arg problem_path)" />
    <!-- conditional planning flags -->
    <param name="use_unknowns" value="false" />
  </node>

  <!-- problem generation -->
  <include file="$(find rosplan_planning_system)/launch/includes/problem_interface.launch">
    <arg name="knowledge_base"   value="rosplan_knowledge_base" />
    <arg name="domain_path"      value="$(arg domain_path)" />
    <!-- problem_path: pddl problem will be automatically generated and placed in this location -->
    <arg name="problem_path"     value="$(arg autom_gen_problem_path)" />
    <arg name="problem_topic"    value="problem_instance" />
  </include>

  <!-- planner interface -->
  <include file="$(find rosplan_planning_system)/launch/includes/planner_interface.launch">
    <arg name="use_problem_topic"    value="$(arg use_problem_topic)" />
    <arg name="problem_topic"        value="/rosplan_problem_interface/problem_instance" />
    <arg name="planner_topic"        value="planner_output" />
    <arg name="domain_path"          value="$(arg domain_path)" />
    <arg name="problem_path"         value="$(arg autom_gen_problem_path)" />
    <arg name="data_path"            value="$(arg data_path)" />
    <arg name="planner_command"      value="timeout 60 $(find rosplan_planning_system)/common/bin/popf DOMAIN PROBLEM" />
  </include>

  <!-- plan parsing -->
  <include file="$(find rosplan_planning_system)/launch/includes/parsing_interface.launch">
    <arg name="knowledge_base"   value="rosplan_knowledge_base" />
    <arg name="planner_topic"    value="/rosplan_planner_interface/planner_output" />
    <arg name="plan_topic"       value="complete_plan" />
  </include>

  <!-- plan dispatch -->
  <include file="$(find rosplan_planning_system)/launch/includes/dispatch_interface.launch">
    <arg name="knowledge_base"           value="rosplan_knowledge_base" />
    <arg name="plan_topic"               value="/rosplan_parsing_interface/complete_plan" />
    <arg name="action_dispatch_topic"    value="action_dispatch" />
    <arg name="action_feedback_topic"    value="action_feedback" />
  </include>

  <node name="mission_plan_publisher" pkg="mimree_executive" type="plan_publisher.py" />

</launch>
